src/views/common/Animation/AnimationUtilities/useTimeSyncing.ts

```diff
@@ -0,0 +1,122 @@
+import React, { useCallback, useMemo, useRef } from "react"
+import useBinarySearchTree, { BstSearchFn, BstSearchResult } from "util/bst"
+import { DebounceThrottleResolver, DebounceThrottleUpdater, useThrottler } from "util/rateLimiters"
+import { AnimationState, AnimationStateAction } from "views/common/Animation/AnimationStateReducer"
+
+
+type TimeLookupFn = (time: number) => BstSearchResult<number> | undefined
+type debounceUpdateRefs = { targetExternalTimeRef: React.MutableRefObject<number | undefined> }
+type debounceUpdateProps = { currentTime: number | undefined }
+type debounceUpdateResolverProps = { committer: (time: number | undefined, setTimeFocus: timeSetter) => void, setterFn: timeSetter }
+type timeSetter = (time: number, o?: any) => void
+
+
+const timeComparison = (a: number, b: number) => a - b
+const throttleRateMs = 100
+
+
+const snapTimeToGrid = (time: number, searchFn: BstSearchFn<number>) => {
+    const nearestTime = searchFn(time)
+    return nearestTime
+}
+
+
+export const syncOutsideTimeToFrameTime = (animationCurrentTime: number | undefined, setOutsideTime: (time: number, o: {autoScrollVisibleTimeRange?: boolean}) => void) => {
+    const currentTime = animationCurrentTime
+    if (currentTime === undefined) return
+    setOutsideTime(currentTime, {autoScrollVisibleTimeRange: true})
+}
+
+
+export const syncFrameToOutsideTime = (outsideTime: number | undefined, findNearestTime: TimeLookupFn, animationStateDispatch: React.Dispatch<AnimationStateAction<any>>) => {
+    if (outsideTime === undefined) return
+    const matchingIndex = findNearestTime(outsideTime)?.baseListIndex
+    if (matchingIndex === undefined) return
+    animationStateDispatch({
+        type: 'SET_CURRENT_FRAME',
+        newIndex: matchingIndex
+    })
+}
+
+
+export const useTimeLookupFn = <T, >(animationState: AnimationState<T>, getTimeFromFrame: (frame: T) => number) => {
+    // TODO: Is this updating too often?
+    const realizedTimestamps = useMemo(() => animationState.frameData.map(d => getTimeFromFrame(d)), [animationState.frameData, getTimeFromFrame])
+    const timeSearchFn = useBinarySearchTree<number>(realizedTimestamps, timeComparison) // do not use an anonymous fn here--results in constant refreshes
+    const findNearestTime = useCallback((time: number) => {
+        return snapTimeToGrid(time, timeSearchFn)
+    }, [timeSearchFn])
+    
+    return findNearestTime
+}
+
+
+type liveSyncFrameToOutsideTimeProps = {
+    outsideTime: number | undefined
+    checkNewTimeWithinEpsilon: (focusTime: number | undefined, replayRateMultiplier: number) => boolean
+    multiplier: number
+    cancelPendingUpdateOfOutsideTime: () => void
+    findNearestTime: (time: number) => BstSearchResult<number> | undefined
+    dispatch: React.Dispatch<AnimationStateAction<any>>
+}
+export const liveSyncFrameToOutsideTime = (props: liveSyncFrameToOutsideTimeProps) => {
```

Comment:
This syncs the frame (i.e. the playback time code) to the time from the external source (i.e. the focus time from the recording context), IFF those times are more than some epsilon value apart.

If you don't do this, you get an infinite loop: the focus time will update to match the playback time, but that's outdated by the time the callback fires.

The epsilon value is in turn set by using 1.5 times the playback rate times the throttling rate: if we run the update once every 100 ms, we expect the new time to be within 100 ms of the old focus time. Only I use 1.5 times that, since things slip. And if we're playing at 10x speed, then it'd be 1.5 s instead of 150 ms. etc.

Reply:
Sorry, to expand on that thought a little:

This is related to ensuring that automatic scrolling of the focus time (in the raster plot) works during active playback of the animation, while still letting a click in the raster plot (or on the scrub bar) reset the time.

There are two times: there's the outside time (here the focus time in the recording context--the one you could set by e.g. clicking the raster plot) and the time that's endogenous to the animation in active playback (call it the "frame time" as it's the time associated with the frame currently being displayed, which is updating ~60 times a second). The tricky part is that neither one is the source of truth: we want the selected frame to update if the user clicks to change the focus, but we also want the animation (during playback) to update the focus time to match the current frame.

This can produce a cycle: the frame animates to the next frame, which updates the frame time, which tries to update the focus time, which triggers an update that resets the frame time. It's all due to the hook that resets the current frame to match the focus time--that hook can only fire when the focus time updates, but there's nothing to distinguish a focus time change that should be honored (due to raster plot click) vs. a focus time that's a lagging effect of the animation updating the focus time.

We got around this in the past by turning off "live" syncing of the focus time: the hook that updates the focus time from the frame time was skipped entirely if playback was active. That way, a manual click (to change the focus time) always works and will update the focus time, triggering an update of the frame time that overwrites the next-frame call; and we just update the focus whenever playback stops (triggering a single focus-time update that won't interfere with the no-longer-ongoing animation of the frames). But this doesn't allow live movement of the focus line in the raster plot.

This only gets worse when you have throttling: the animation will play at one speed, the focus time updates are throttled (for performance reasons) to a lower speed; and as a result the updates to the focus will never quite sync up with the time implied by the current frame, and you can get stuck in further cycles or introduce playback lag.

The way I've implemented it here is to introduce an "epsilon" distance which measures the difference between the current focus time and the time the animation playback wants to update the focus to. Basically, we still have a "reset the frame to the focus" hook that fires on every update of the focus--EXCEPT that hook now knows the time difference between the current focus state and the updated focus state. It also knows how fast playback is happening (based on the multiplier). So if the new focus time is within the playback-rate-adjusted timespan implied by the sync throttling interval, the hook knows not to try to update the frame to match the focus, and just accepts the new focus time. Instead of deferring updates to the focus time until playback has stopped, we just break the potentially cyclic updates by conditionally not updating the frame from the focus.

This works because we know that the frames play back at a certain rate--the baseline playback rate for the animation is that 1 second of timestamp time should take 1 second to play back. So if we update the focus to match the frame with a callback that runs once every 100 ms, we should expect any discrepancy up to 100 ms between the focus and frame time to be due to the animation, so don't update the frame in response to the changed focus. (Or if we're playing at 10x speed, then a 100-ms update interval means we should expect and accept time discrepancies up to 1 sec.) (Also I multiply the window by 1.5 to give a bit of extra slush.) If the focus is updated to something outside that time window--like because the user clicked in the raster plot to change it--the difference is big enough, so it resets the position of the animation as per usual.

The upshot is that live focus updates work, in a throttled way, while still allowing clicks in the raster plot to work during playback.

The function that would nominally do this callback is complicated and most of its parameters (like the playback rate multiplier, the state dispatcher, etc.) aren't really relevant to when it should run, so to make the relationship clearer I've written the update function as a callback that takes the external time (focus time). This gets returned by the `useLiveTimeSyncing` hook as the `handleOutsideTimeUpdate` function returned in l. 115 below. That in turn gets called in a `useEffect` hook in `TrackPositionAnimationView` l. 154--which `useEffect` hook fires only when the focus time is updated.

This was all a bit complicated; I hope that this note explains more than it confuses...